### PR TITLE
Editor refactor

### PR DESCRIPTION
**Pull request by alexkrolick**
- feature: debounce change handling
- feature: show save state in floating loading indicator
- add js tooling (eslint, prettier)

---

- view without whitespace diff: https://github.com/jellypbc/poster/pull/38/files?utf8=✓&diff=split&w=1

### CODE REVIEW

**Comment by dennyluan** [comment]
Nice! I've been meaning to do the debounce fix, this looks great.

Re: `TODO: should prevent closing tab while hasChanges state is true` => was thinking we could dump things into localstorage and let fetch that version if things hadn't changed. Might get complicated with conflicts. At the very least just prompt an alert "Are you sure you want to exit?" modal.

File: app/javascript/components/HtmlEditor.js [hunk 1]
```diff
@@ -1,85 +1,84 @@
...
+		doc => {
+			const { onChange } = this.props
+			onChange(this.serialize(doc))
+		},
+		350,
+		{ maxWait: 1000 }
+	)
+
+	render() {
+		const { autoFocus, options, attributes, render, nodeViews } = this.props
```
Comment (dennyluan) [inline app/javascript/components/HtmlEditor.js]: heh i bring bad ruby habits into my JS

**Comment by dennyluan** [comment]
This cleans up a lot of the weird stuff going in the editor 😅 looks good

**Comment by alexkrolick** [comment]
> Re: TODO: should prevent closing tab while hasChanges state is true => was thinking we could dump things into localstorage and let fetch that version if things hadn't changed. Might get complicated with conflicts. At the very least just prompt an alert "Are you sure you want to exit?" modal

- Yes, just adding an event listener to onbeforeunload will prevent navigation with unsaved changes
- Conflicts require something like a vector clock (simple) or CRDT, can do that later

**Pull request merged**